File: examples/game-runtime/uiux-flow.js
// UI/UX Flow, Onboarding and Accessibility v1 (#1660, contract
// docs/long-form-systems-v1.md).
//
// A deterministic, probe-exposed in-game UI/UX flow for the existing
// game-runtime: declared screens (menu / onboarding / hud / settings),
// deterministic navigation transitions, and declared accessibility options
// (toggles and enums). It is in-game UI in the JS runtime, read-only with
// respect to trusted state — never a Studio trusted-write surface.
//
// The module is pure and deterministic: given the same flow spec and the same
// navigation/accessibility inputs it reproduces the same flow state and the same
// read-only view. Validation fails closed with a clear diagnostic: an undeclared
// screen, a non-deterministic transition (two transitions sharing from+action),
// an unreachable screen, or a missing/invalid accessibility option is rejected.
(() => {
  const SPEC_SCHEMA = 'uiux-flow-v1';
  const STATE_SCHEMA = 'ouroforge.uiux-flow-state.v1';
  const SCREEN_KINDS = ['menu', 'onboarding', 'hud', 'settings'];
  const OPTION_TYPES = ['toggle', 'enum'];
  // Canonical read-only/proposal-only boundary, mirrored from the Rust contract
  // (crates/ouroforge-core/src/uiux_flow.rs). The runtime rejects any other
  // value so a scene cannot load with a misleading UI/UX boundary.
  const BOUNDARY =
    'rust-trusted-state; browser/studio read-only; generation proposal-only via review/apply/trust-gradient';

  function fail(message) {
    throw new Error(`uiux flow spec invalid: ${message}`);
  }

  function isPlainObject(value) {
    return value !== null && typeof value === 'object' && !Array.isArray(value);
  }

  function clone(value) {
    return JSON.parse(JSON.stringify(value));
  }

  function normalizeSpec(spec) {
    if (!isPlainObject(spec)) fail('spec must be an object');
    if (spec.schemaVersion !== SPEC_SCHEMA) {
      fail(`schemaVersion must be ${SPEC_SCHEMA}`);
    }
    if (spec.boundary !== BOUNDARY) {
      fail('boundary must be the canonical read-only/proposal-only contract');
    }

    if (!Array.isArray(spec.screens) || spec.screens.length === 0) {
      fail('screens must be a non-empty array');
    }
    const screens = [];
    const screenIds = new Set();
    for (const screen of spec.screens) {
      if (!isPlainObject(screen)) fail('each screen must be an object');
      const id = typeof screen.id === 'string' ? screen.id.trim() : '';
      if (!id) fail('screen id must be a non-empty string');
      if (screenIds.has(id)) fail(`duplicate screen id ${id}`);
      if (!SCREEN_KINDS.includes(screen.kind)) {
        fail(`screen ${id} has invalid kind ${screen.kind}`);
      }
      screenIds.add(id);
      screens.push({ id, kind: screen.kind });
    }

    const initialScreen = typeof spec.initialScreen === 'string' ? spec.initialScreen : '';
    if (!screenIds.has(initialScreen)) fail('initialScreen must be a declared screen');

    if (!Array.isArray(spec.transitions)) fail('transitions must be an array');
    const transitions = [];
    const transitionMap = {};
    for (const transition of spec.transitions) {
      if (!isPlainObject(transition)) fail('each transition must be an object');
      const from = typeof transition.from === 'string' ? transition.from : '';
      const action = typeof transition.action === 'string' ? transition.action.trim() : '';
      const to = typeof transition.to === 'string' ? transition.to : '';
      if (!screenIds.has(from)) fail(`transition references undeclared from-screen ${from}`);
      if (!screenIds.has(to)) fail(`transition references undeclared to-screen ${to}`);
      if (!action) fail('transition action must be a non-empty string');
      if (!transitionMap[from]) transitionMap[from] = {};
      if (Object.prototype.hasOwnProperty.call(transitionMap[from], action)) {
        fail(`non-deterministic transition: ${from} + ${action}`);
      }
      transitionMap[from][action] = to;
      transitions.push({ from, action, to });
    }

    // Reachability: every declared screen must be reachable from the initial
    // screen, otherwise the flow has a dead UI state.
    const reachable = new Set([initialScreen]);
    const queue = [initialScreen];
    while (queue.length > 0) {
      const current = queue.shift();
      const outgoing = transitionMap[current] || {};
      for (const action of Object.keys(outgoing)) {
        const to = outgoing[action];
        if (!reachable.has(to)) {
          reachable.add(to);
          queue.push(to);
        }
      }
    }
    for (const screen of screens) {
      if (!reachable.has(screen.id)) fail(`screen ${screen.id} is unreachable from initialScreen`);
    }

    if (!Array.isArray(spec.accessibilityOptions) || spec.accessibilityOptions.length === 0) {
      fail('accessibilityOptions must declare at least one option');
    }
    const options = [];
    const optionIds = new Set();
    for (const option of spec.accessibilityOptions) {
      if (!isPlainObject(option)) fail('each accessibility option must be an object');
      const id = typeof option.id === 'string' ? option.id.trim() : '';
      if (!id) fail('accessibility option id must be a non-empty string');
      if (optionIds.has(id)) fail(`duplicate accessibility option ${id}`);
      if (!OPTION_TYPES.includes(option.type)) {
        fail(`accessibility option ${id} has invalid type ${option.type}`);
      }
      if (option.type === 'toggle') {
        if (typeof option.default !== 'boolean') {
          fail(`toggle option ${id} default must be a boolean`);
        }
        options.push({ id, type: 'toggle', default: option.default });
      } else {
        if (!Array.isArray(option.values) || option.values.length === 0) {
          fail(`enum option ${id} must declare a non-empty values array`);
        }
        const values = option.values.map(String);
        if (!values.includes(String(option.default))) {
          fail(`enum option ${id} default must be one of its values`);
        }
        options.push({ id, type: 'enum', default: String(option.default), values });
      }
      optionIds.add(id);
    }

    const accessibility = {};
    for (const option of options) accessibility[option.id] = option.default;

    return {
      schema: STATE_SCHEMA,
      screens,
      transitionMap,
      options,
      currentScreen: initialScreen,
      visited: [initialScreen],
      accessibility,
      lastNavigation: null,
      boundary: BOUNDARY,
    };
  }

  function screenKind(state, id) {
    const screen = state.screens.find((entry) => entry.id === id);
    return screen ? screen.kind : null;
  }

  // Apply one navigation action. A declared (currentScreen, action) transition
  // moves the flow; an undeclared action is a deterministic no-op recorded as a
  // rejected navigation (in-game UI ignores invalid navigation rather than
  // failing the runtime).
  function navigate(state, action) {
    const next = clone(state);
    const act = typeof action === 'string' ? action.trim() : '';
    const outgoing = state.transitionMap[state.currentScreen] || {};
    if (act && Object.prototype.hasOwnProperty.call(outgoing, act)) {
      next.currentScreen = outgoing[act];
      if (!next.visited.includes(next.currentScreen)) next.visited.push(next.currentScreen);
      next.lastNavigation = {
        action: act,
        from: state.currentScreen,
        to: next.currentScreen,
        accepted: true,
      };
    } else {
      next.lastNavigation = {
        action: act,
        from: state.currentScreen,
        to: state.currentScreen,
        accepted: false,
      };
    }
    return next;
  }

  // Set a declared accessibility option. An undeclared option or an
  // out-of-domain value fails closed.
  function setAccessibility(state, optionId, value) {
    const option = state.options.find((entry) => entry.id === optionId);
    if (!option) throw new Error(`uiux flow: unknown accessibility option ${optionId}`);
    const next = clone(state);
    if (option.type === 'toggle') {
      if (typeof value !== 'boolean') {
        throw new Error(`uiux flow: toggle option ${optionId} requires a boolean`);
      }
      next.accessibility[optionId] = value;
    } else {
      const candidate = String(value);
      if (!option.values.includes(candidate)) {
        throw new Error(`uiux flow: enum option ${optionId} value ${candidate} is not allowed`);
      }
      next.accessibility[optionId] = candidate;
    }
    return next;
  }

  function onboardingComplete(state) {
    const visitedOnboarding = state.visited.some((id) => screenKind(state, id) === 'onboarding');
    const currentlyOnboarding = screenKind(state, state.currentScreen) === 'onboarding';
    return visitedOnboarding && !currentlyOnboarding;
  }

  // Read-only world-state view exposed through the runtime probe.
  function worldStateView(state) {
    return {
      schema: STATE_SCHEMA,
      currentScreen: state.currentScreen,
      currentKind: screenKind(state, state.currentScreen),
      visitedScreens: clone(state.visited),
      screenCount: state.screens.length,
      accessibility: clone(state.accessibility),
      accessibilityOptionCount: state.options.length,
      onboardingComplete: onboardingComplete(state),
      lastNavigation: clone(state.lastNavigation),
      boundary: state.boundary,
    };
  }

  const api = {
    SPEC_SCHEMA,
    STATE_SCHEMA,
    normalizeSpec,
    createState: normalizeSpec,
    navigate,
    setAccessibility,
    worldStateView,
  };

  if (typeof window !== 'undefined') {
    window.OuroforgeUiuxFlow = api;
  }
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api;
  }
})();

File: crates/ouroforge-core/src/uiux_flow.rs
//! UI/UX Flow, Onboarding and Accessibility v1 (#1660) — the trusted contract
//! for the in-game UI/UX flow that the JS runtime implements and exposes through
//! the read-only probe.
//!
//! This module owns the trusted validation of a UI/UX flow contract: declared
//! screens (menu / onboarding / hud / settings), deterministic navigation
//! transitions, and declared accessibility options (toggles and enums). It adds
//! no runtime; the in-game UI itself is the deterministic JavaScript runtime
//! (`examples/game-runtime/uiux-flow.js`), read-only with respect to trusted
//! state. This Rust contract is the trusted, machine-checkable acceptance shape
//! the runtime flow must satisfy and the mirror of the runtime contract test.
//!
//! Fail-closed: validation rejects an undeclared screen, an undeclared
//! `initialScreen`, a non-deterministic transition (two transitions sharing
//! from+action), an unreachable screen, a missing accessibility set, an invalid
//! option type/default, and a non-canonical boundary — each with a structured
//! reason. It makes no UX-quality or taste judgement; it only checks that the
//! declared flow is deterministic, fully reachable, and carries the declared
//! accessibility affordances.
//!
//! Boundary: Rust/local owns this trusted contract. The in-game UI is read-only
//! JavaScript runtime presentation; the browser never writes trusted state.
//! See `docs/long-form-systems-v1.md`.

use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet, VecDeque};

pub const UIUX_FLOW_SCHEMA_VERSION: &str = "uiux-flow-v1";

/// Canonical trust boundary recorded on every flow contract.
pub const UIUX_FLOW_BOUNDARY: &str =
    "rust-trusted-state; browser/studio read-only; generation proposal-only via review/apply/trust-gradient";

/// The declared kinds of UI/UX screen.
pub const UIUX_SCREEN_KINDS: &[&str] = &["menu", "onboarding", "hud", "settings"];

/// The declared kinds of accessibility option.
pub const UIUX_OPTION_TYPES: &[&str] = &["toggle", "enum"];

/// One declared screen.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UiuxScreen {
    pub id: String,
    pub kind: String,
}

/// One declared navigation transition.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UiuxTransition {
    pub from: String,
    pub action: String,
    pub to: String,
}

/// One declared accessibility option. `default` is a boolean for a `toggle` and
/// a string for an `enum`; `values` is present only for an `enum`.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UiuxAccessibilityOption {
    pub id: String,
    #[serde(rename = "type")]
    pub kind: String,
    pub default: serde_json::Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub values: Option<Vec<String>>,
}

/// The declarative UI/UX flow contract.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UiuxFlowContract {
    pub schema_version: String,
    pub boundary: String,
    pub screens: Vec<UiuxScreen>,
    pub initial_screen: String,
    pub transitions: Vec<UiuxTransition>,
    pub accessibility_options: Vec<UiuxAccessibilityOption>,
}

/// Read-only summary of a validated flow contract for browser/Studio surfaces.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UiuxFlowReadModel {
    pub schema_version: String,
    pub initial_screen: String,
    pub screen_count: usize,
    pub transition_count: usize,
    pub accessibility_option_count: usize,
    pub screen_kinds: Vec<String>,
    pub boundary: String,
}

impl UiuxFlowContract {
    /// Parse and validate a flow contract from JSON, failing closed on malformed
    /// or out-of-contract input.
    pub fn from_json_str(input: &str) -> Result<Self> {
        let contract: Self = serde_json::from_str(input)
            .map_err(|err| anyhow!("invalid uiux flow contract json: {err}"))?;
        contract.validate()?;
        Ok(contract)
    }

    /// Validate the flow contract, failing closed on any contract violation.
    pub fn validate(&self) -> Result<()> {
        if self.schema_version != UIUX_FLOW_SCHEMA_VERSION {
            return Err(anyhow!(
                "unexpected uiux flow schema version: {} (expected {})",
                self.schema_version,
                UIUX_FLOW_SCHEMA_VERSION
            ));
        }
        if self.boundary != UIUX_FLOW_BOUNDARY {
            return Err(anyhow!(
                "uiux flow boundary must be the canonical read-only/proposal-only contract"
            ));
        }

        if self.screens.is_empty() {
            return Err(anyhow!("uiux flow must declare at least one screen"));
        }
        let mut screen_ids: BTreeSet<&str> = BTreeSet::new();
        for screen in &self.screens {
            if screen.id.trim().is_empty() {
                return Err(anyhow!("uiux screen id must not be empty"));
            }
            if !UIUX_SCREEN_KINDS.contains(&screen.kind.as_str()) {
                return Err(anyhow!(
                    "uiux screen '{}' has invalid kind '{}'",
                    screen.id,
                    screen.kind
                ));
            }
            if !screen_ids.insert(screen.id.as_str()) {
                return Err(anyhow!("duplicate uiux screen id '{}'", screen.id));
            }
        }

        if !screen_ids.contains(self.initial_screen.as_str()) {
            return Err(anyhow!(
                "uiux initialScreen '{}' is not a declared screen",
                self.initial_screen
            ));
        }

        // Deterministic transitions: at most one transition per (from, action).
        let mut transition_map: BTreeMap<&str, BTreeMap<&str, &str>> = BTreeMap::new();
        for transition in &self.transitions {
            if !screen_ids.contains(transition.from.as_str()) {
                return Err(anyhow!(
                    "uiux transition references undeclared from-screen '{}'",
                    transition.from
                ));
            }
            if !screen_ids.contains(transition.to.as_str()) {
                return Err(anyhow!(
                    "uiux transition references undeclared to-screen '{}'",
                    transition.to
                ));
            }
            if transition.action.trim().is_empty() {
                return Err(anyhow!("uiux transition action must not be empty"));
            }
            let outgoing = transition_map.entry(transition.from.as_str()).or_default();
            if outgoing
                .insert(transition.action.as_str(), transition.to.as_str())
                .is_some()
            {
                return Err(anyhow!(
                    "non-deterministic uiux transition: '{}' + '{}'",
                    transition.from,
                    transition.action
                ));
            }
        }

        // Reachability: every screen reachable from the initial screen.
        let mut reachable: BTreeSet<&str> = BTreeSet::new();
        reachable.insert(self.initial_screen.as_str());
        let mut queue: VecDeque<&str> = VecDeque::new();
        queue.push_back(self.initial_screen.as_str());
        while let Some(current) = queue.pop_front() {
            if let Some(outgoing) = transition_map.get(current) {
                for to in outgoing.values() {
                    if reachable.insert(to) {
                        queue.push_back(to);
                    }
                }
            }
        }
        for screen in &self.screens {
            if !reachable.contains(screen.id.as_str()) {
                return Err(anyhow!(
                    "uiux screen '{}' is unreachable from initialScreen",
                    screen.id
                ));
            }
        }

        // Accessibility affordances must be present and well-formed.
        if self.accessibility_options.is_empty() {
            return Err(anyhow!(
                "uiux flow must declare at least one accessibility option"
            ));
        }
        let mut option_ids: BTreeSet<&str> = BTreeSet::new();
        for option in &self.accessibility_options {
            if option.id.trim().is_empty() {
                return Err(anyhow!("uiux accessibility option id must not be empty"));
            }
            if !option_ids.insert(option.id.as_str()) {
                return Err(anyhow!(
                    "duplicate uiux accessibility option '{}'",
                    option.id
                ));
            }
            if !UIUX_OPTION_TYPES.contains(&option.kind.as_str()) {
                return Err(anyhow!(
                    "uiux accessibility option '{}' has invalid type '{}'",
                    option.id,
                    option.kind
                ));
            }
            match option.kind.as_str() {
                "toggle" => {
                    if !option.default.is_boolean() {
                        return Err(anyhow!(
                            "uiux toggle option '{}' default must be a boolean",
                            option.id
                        ));
                    }
                    if option.values.is_some() {
                        return Err(anyhow!(
                            "uiux toggle option '{}' must not declare enum values",
                            option.id
                        ));
                    }
                }
                "enum" => {
                    let values = option.values.as_ref().ok_or_else(|| {
                        anyhow!("uiux enum option '{}' must declare values", option.id)
                    })?;
                    if values.is_empty() {
                        return Err(anyhow!(
                            "uiux enum option '{}' must declare a non-empty values array",
                            option.id
                        ));
                    }
                    let default = option.default.as_str().ok_or_else(|| {
                        anyhow!("uiux enum option '{}' default must be a string", option.id)
                    })?;
                    if !values.iter().any(|v| v == default) {
                        return Err(anyhow!(
                            "uiux enum option '{}' default must be one of its values",
                            option.id
                        ));
                    }
                }
                _ => unreachable!("option kind already validated"),
            }
        }

        Ok(())
    }

    /// Derive the read-only presentation summary for browser/Studio surfaces.
    pub fn read_model(&self) -> UiuxFlowReadModel {
        UiuxFlowReadModel {
            schema_version: self.schema_version.clone(),
            initial_screen: self.initial_screen.clone(),
            screen_count: self.screens.len(),
            transition_count: self.transitions.len(),
            accessibility_option_count: self.accessibility_options.len(),
            screen_kinds: self.screens.iter().map(|s| s.kind.clone()).collect(),
            boundary: self.boundary.clone(),
        }
    }
}
